frontend/src/components/UserProfileContext.tsx

```diff
@@ -0,0 +1,50 @@
+import { useAuth0 } from '@auth0/auth0-react'
+import { createContext, FunctionComponent, useEffect, useState } from 'react'
+
+export interface UserProfile {
+  id: number
+  isAdmin: boolean
+}
+
+const fetchProfile = (token: string) => {
+  console.log("fetching the user's profile")
+
+  return fetch('/profile', {
+    headers: { Authorization: `Bearer ${token}` },
+  })
+}
+
+const UserProfileContext = createContext<UserProfile | undefined>(undefined)
+
+const UserProfileProvider: FunctionComponent = ({ children }) => {
+  const { isAuthenticated, getAccessTokenSilently } = useAuth0()
+  const [profileIsLoaded, setProfileIsLoaded] = useState(false)
+  const [profile, setProfile] = useState<UserProfile>()
+
+  useEffect(() => {
+    if (isAuthenticated && !profileIsLoaded) {
+      getAccessTokenSilently()
+        .then(fetchProfile)
+        .then((response) => {
+          if (response.ok) {
+            setProfileIsLoaded(true)
+          } else {
+            console.error('Non-OK server response retrieving profile')
+          }
+
+          return response.json()
+        })
+        .then((data) => {
+          setProfile(data)
+        })
+    }
+  }, [isAuthenticated, getAccessTokenSilently, profileIsLoaded])
```

Comment:
This code was copied over from AppRoot, except that we're not storing the user's profile. This way, any component nested within the provider can access that information without an extra call to the backend.

Reply:
Nice, yet another react feature for me to learn 😄